src/rust/ide/src/controller/module.rs

```diff
@@ -170,6 +203,23 @@ shared! { Handle
         pub fn subscribe_graph_notifications(&mut self) -> Subscriber<notification::Graphs> {
             self.graph_notifications.subscribe()
         }
+
+        /// Returns metadata for given node, if present.
+        pub fn node_metadata(&mut self, id:ast::ID) -> FallibleResult<NodeMetadata> {
+            self.module.metadata.ide.node.get(&id).cloned()
+                .ok_or_else(|| NodeMetadataNotFound(id).into())
+        }
+
+        /// Modify metadata of given node.
+        /// If ID doesn't have metadata, empty (default) metadata is inserted.
+        pub fn with_node_metadata(&mut self, id:ast::ID, fun:impl FnOnce(&mut NodeMetadata)) {
+            fun(self.module.metadata.ide.node.entry(id).or_default());
```

Comment:
There is a little danger here: we are calling _some_ callback while having borrowed module controller state. So the user cannot make any operation which potentially may borrow the same module controller again. Usually this is hard to guarantee (we should check all calls for "with_node_metadata").

I propose to implement this function in Handle implementation (not inside shared! macro) and do it this way:
1. borrow data take the metadata entry from the map.
2. without borrow call the callback.
3. borrow and put the modified ast back

Reply:
Good point @farmaazon  👍